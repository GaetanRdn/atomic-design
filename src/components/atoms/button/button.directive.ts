import { Directive, Input } from '@angular/core';

@Directive({
  selector: 'button[adButton]',
  host: {
    class: 'ad-button',
    '[class.ad-small]': 'size === "small"',
    '[class.ad-medium]': 'size === "medium"',
    '[class.ad-large]': 'size === "large"',
    '[class.outlined]': 'outlined',
    '[class.ad-primary]': 'color === "primary"',
    '[class.ad-accent]': 'color === "accent"',
    '[class.ad-warn]': 'color === "warn"'
  }
})
export class ButtonDirective {
  @Input()
  public size: 'small' | 'medium' | 'large' = 'medium';

  @Input()
  public outlined: boolean = false;

  @Input()
  public color?: 'primary' | 'accent' | 'warn';
}
