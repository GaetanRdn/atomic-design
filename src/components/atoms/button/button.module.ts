import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonDirective } from 'src/components/atoms/button/button.directive';



@NgModule({
  declarations: [
    ButtonDirective
  ],
  imports: [
    CommonModule
  ]
})
export class ButtonModule { }
