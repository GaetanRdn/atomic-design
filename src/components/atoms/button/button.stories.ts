import { CommonModule } from '@angular/common';
import { Component, Input } from '@angular/core';
import { Meta, Story } from '@storybook/angular';
import { ButtonDirective } from 'src/components/atoms/button/button.directive';

@Component({
  selector: 'adr-dummy',
  template: `
      <button adrButton [size]="size" [outlined]="outlined" [color]="color">Click</button>`
})
class DummyComponent {
  @Input()
  public size: 'small' | 'medium' | 'large' = 'medium';

  @Input()
  public outlined: boolean = false;

  @Input()
  public color: 'primary' | 'accent' | 'warn' = 'primary';
}

export default {
  title: 'atoms/button',
  component: DummyComponent
} as Meta;

const Template: Story<DummyComponent> = (args) => ({
  props: args, moduleMetadata: { declarations: [ButtonDirective], imports: [CommonModule] }
});

export const Default = Template.bind({});

