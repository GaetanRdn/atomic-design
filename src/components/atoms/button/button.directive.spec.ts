import { Component } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ButtonDirective } from './button.directive';

describe('ButtonDirective', () => {
  let fixture: ComponentFixture<HostComponent>;

  beforeEach(() => {
    fixture = TestBed.configureTestingModule({
      declarations: [ButtonDirective, HostComponent]
    }).createComponent(HostComponent);
  });

  test('should create default button', async () => {
    // WHEN
    fixture.detectChanges();
    await fixture.whenRenderingDone();

    // THEN
    expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
  });

  test.each([
    'small',
    'medium',
    'large'
  ] as ('small' | 'medium' | 'large')[])('when size is %s then check snapshot', (size: 'small' | 'medium' | 'large') => {
    // GIVEN
    fixture.componentInstance.size = size;

    // WHEN
    fixture.detectChanges();

    // THEN
    expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
  });

  test.each([true, false])('when outlined is set to %s then check snapshot', (outlined: boolean) => {
    // GIVEN
    fixture.componentInstance.outlined = outlined;

    // WHEN
    fixture.detectChanges();

    // THEN
    expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
  });

  test.each([
    'primary',
    'accent',
    'warn'
  ] as ('primary' | 'accent' | 'warn')[])('when color is %s then check snapshot', (color: 'primary' | 'accent' | 'warn') => {
    // GIVEN
    fixture.componentInstance.color = color;

    // WHEN
    fixture.detectChanges();

    // THEN
    expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
  });
});

@Component({
  template: `
      <button adButton [size]="size" [outlined]="outlined" [color]="color">Click Me</button>`
})
class HostComponent {
  public size?: 'small' | 'medium' | 'large';

  public outlined?: boolean;

  public color?: 'primary' | 'accent' | 'warn';
}
