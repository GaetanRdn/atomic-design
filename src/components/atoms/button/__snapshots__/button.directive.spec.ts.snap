// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDirective should create default button 1`] = `
<button
  adbutton=""
  class="ad-button"
>
  Click Me
</button>
`;

exports[`ButtonDirective when color is accent then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button ad-accent"
>
  Click Me
</button>
`;

exports[`ButtonDirective when color is primary then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button ad-primary"
>
  Click Me
</button>
`;

exports[`ButtonDirective when color is warn then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button ad-warn"
>
  Click Me
</button>
`;

exports[`ButtonDirective when outlined is set to false then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button"
>
  Click Me
</button>
`;

exports[`ButtonDirective when outlined is set to true then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button outlined"
>
  Click Me
</button>
`;

exports[`ButtonDirective when size is large then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button ad-large"
>
  Click Me
</button>
`;

exports[`ButtonDirective when size is medium then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button ad-medium"
>
  Click Me
</button>
`;

exports[`ButtonDirective when size is small then check snapshot 1`] = `
<button
  adbutton=""
  class="ad-button ad-small"
>
  Click Me
</button>
`;
